### ISSUE TITLE
returning a pointer to a struct from the FFI wraps it automatically in a struct

### ISSUE THREAD

**Issue by jonenst** [slavapestov/factor#1522]
The docs say:
> If a C function is declared as returning a pointer to a struct, it will return an alien instance. This is because there is no way to distinguish between a pointer to a single struct and a pointer to an array of zero or more structs. It is up to the caller to wrap it in a struct, or a specialized array of structs, respectively.
But right now it returns a struct.

I think that the docs makes sense, because right now we have to access the ```(underlying)``` slot of the struct to get the address and make a <direct-T-array> from it. It's undocumented and private.

I was thinking maybe we can be smart and return a struct if the SPECIALIZED-ARRAY: has not been declared, but this is bad because changing something in one place of a source file (adding specialized-array) breaks a function that can be at the other end of the file..

Another solution might be to introduce new syntax to tell what to convert to ? like
```
FUNCTION: mystruct* foo ( ) ! converts to S{ foo }
FUNCTION: mystruct[] foo ( ) ! converts to foo-array{ ~foo~ ~foo~ }
``` ?

**Comment by bjourne** [comment]
Returning arrays of structs is a very uncommon use case so I don't think it warrants special syntax. 

But the docs are wrong as you say. Both returning a struct and returning a struct pointer will push a struct instance on the stack. The only user visible difference is that in the former case, the struct is allocated as a byte array in memory inside the Factor heap. In the latter case, the struct exists in memory outside the factor heap and is represented by an alien instance.

**Comment by jonenst** [comment]
You are right about the difference between ```FUNCTION: mystruct``` and ```FUNCTION: mystruct*```, but this is not the problem I was describing. Also the caller of the function can just look at the function definition and see what the return type is..

The problem I was talking about is that 2 different C functions (one returning a struct by pointer and one return an array of structs) that must have the same FFI definition (```FUNCTION: mystruct*```).


So do you think we should revert the behavior to the one described in the docs ?
When always returning a struct, although it's the good thing to do most of the time, in some cases you have to fetch the ```(underlying)``` and use that as an array.

**Comment by bjourne** [comment]
If it worked like the docs specify, you would have to write `some-func mystruct memory>struct` in a billion places. That is not practical, even if it is often better to be explicit than implicit. 

For a function that returns an array of structs, you always want to declare it with `void*` as the return type. Then you cast it to the desired array: `some-func 3 <direct-mystruct-array>`. C language arrays doesn't tell you how long they are so you must always specify their size. That's why special syntax that unpacks arrays is tricky.

**Comment by jonenst** [comment]
Well we could have something like
```
FUNCTION: mystruct[ <arbitrary code here> ] foo ( int* len )
```
where the arbitrary code would have a stack-effect of ( ... -- ... length ) and be called after the C function returns with the same arguments (ie ```{ [ c-func ] [ <arbitrary code here> ] } N ncleave``` ). it could even be executed in a local scope where the arguments of the C function are bound to lexical variables with their names from the C declaration ( like [| | .. ] ). maybe even add dereferenced versions of pointers arguments as alternate lexical variables, which would allow this:
```
! Might get messy if most arguments need to be dropd
FUNCTION: mystruct[ drop nip int deref ] foo ( int arg1, int *len, int arg2 ) 
! Instead, use lexicals variables ?
FUNCTION: mystruct[ *len int deref ] foo ( int arg1, int arg2, int *len )
! *len would still be available but most of the time we just want len ?
FUNCTION: mystruct[ len ] foo ( int arg1, int arg2, int *len )
```
Note: this could work for other pointers for functions returning arrays of ints, floats, etc..
Note2: It's fun that arrays in C use square brackets to specify the size as in int[10], and in factor square brackets are for the arbitrary code quotations. So in ```FUNCTION: mystrct[ <code> ] foo ()``` we have both of them. I'm not sure if that's making it better or just more confusing.

Anyway, on the short term (and maybe even on the long term, as the syntax above is somehow complex and may not be worth it) declaring it as void* is a good workaround.  As you said functions returning an array of struct are rare.
So let's just change the docs to say:
> If a C FFI function is declared in factor as returning a pointer to a struct, it will return a struct instance backed by the memory at the pointer's value. This works correctly with C functions returning a struct by reference. However, this forces C functions that are returning an array of structs to be declared in factor to return a generic void* pointer. The caller must then wrap it in a specialized array of structs. This is because there is no way to distinguish between a pointer to a single struct and a pointer to an array of zero or more structs. C FFI functions declared as returning a pointer to a struct behave this way because returning by reference is more frequent than returning an array, and because when returning an array, the caller must supply the length of the array, which might need arbitrary code and so makes the declaration complex anyway.

What do you think ?

**Comment by bjourne** [comment]
That text looks good to me!
